=act improve ByteString#dropRight(...)

monkey-mas opened the pull request:
Currently, we recursively call `dropRight(...)` and obtain the rest of `bytestrings` by calling `Vector#init`, which would not be quite efficient.

Instead of doing this, we can do as follows so that we can achieve better performance:
1. Seek the index of _last_ vector element we need to _dropRight_
2. Find the number of characters left to _dropRight_ from the _last_ ByteString1 element.
3. Create `ByteString` based on the information we obtained from 1 and 2

Then we just need to create a new `Vector[ByteString1]` at most twice, which should be better than the current implementation, i.e., create the rest `bytestrings` every time we check `bytestrings(Vector[ByteString1])` element, which ends up O(N) _init_ execution where _N_ is the length of `bytestrings`.
### Benchmark

I measured the performance using sbt-jmh with this [code](https://gist.github.com/monkey-mas/684ec6a5e2a0ce42f28417dcd53c9636).  Benchmark code was executed with some parameters such as `-t1 -f2 -wi 10 -i 20`.
#### Test environment

```
OS: Ubuntu 14.0.4 (Linux kernel version 3.13.0-95-generic)
CPU: Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
VM version: JDK 1.8.0_77
```
#### Before

```
Benchmark                                  Mode  Cnt           Score         Error  Units
DropRight.dropRight_avg                   thrpt   40       26184.435 ±     403.058  ops/s
DropRight.dropRight_best                  thrpt   40     8707856.806 ±   50699.447  ops/s
DropRight.dropRight_greater_or_eq_to_len  thrpt   40       12289.742 ±     149.310  ops/s
DropRight.dropRight_negative              thrpt   40  1191630555.447 ± 5225134.083  ops/s
DropRight.dropRight_worst                 thrpt   40       12329.187 ±     283.525  ops/s
```
#### After

```
DropRight.dropRight_avg                   thrpt   40      630595.880 ±     3895.122  ops/s
DropRight.dropRight_best                  thrpt   40     7608841.379 ±   264211.274  ops/s
DropRight.dropRight_greater_or_eq_to_len  thrpt   40   877987542.956 ± 12697839.317  ops/s
DropRight.dropRight_negative              thrpt   40  1125036409.900 ± 17180858.452  ops/s
DropRight.dropRight_worst                 thrpt   40      224135.469 ±     4563.056  ops/s
```

monkey-mas committed:
=act improve ByteString#dropRight(...)

Currently, we recursively call dropRight(...) obtaining the rest of bytestrings by calling Vector#init,
which would not be quite efficient.

Instead of doing this, we can do as follows so that we can achieve better performance:
1. Seek the index of _last_ vector element we need to _dropRight_
2. Find the number of characters left to dropRight from the _last_ ByteString1 element.
3. Create ByteString based on the information we obtained from 1 and 2

Then we just need to create a new Vector[ByteString1] at most twice, which should be
better than the current implementation, i.e., create the rest bytestrings every time we check
bytestrings(Vector[ByteString1]) element, which ends up O(N) _init_ execution where _N_ is
the length of bytestrings.

johanandren commented:
OK TO TEST

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -448,20 +448,23 @@ object ByteString {
+      else if (n >= length) ByteString.empty
+      else dropRight0(n)
+
+    private[akka] def dropRight0(n: Int): ByteString = {
```
I'd really prefer hiding those methods completely, can be `private` (or hidden inside the method itself)

monkey-mas replied:
Thanks for pointing it out!!! Yes, we should hide it :)
I'm planning to use `dropRight(...)` or `take(...)` to improve `slice(...)` so I might make `take0(...)`  `private` (instead of _completely_ hiding it inside `take(...)`)
Anyway will fix it for sure :+1:

ktoso commented:
LGTM, thanks a lot for addressing this.
dropRight wasn't on hot-paths we cared about so indeed it was implemented a bit more leisurely, good to get it up to speed as well :)

Would you mind hiding the dropRight0 please?

patriknw commented:
LGTM

monkey-mas committed:
add benchmark of ByteString#dropRight(...)

monkey-mas commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -448,20 +448,23 @@ object ByteString {
+
+      if (last == 0) bytestrings(0).dropRight(restToDropRight)
+      else if (restToDropRight == 0) new ByteStrings(bytestrings.take(last + 1), length - n)
+      else new ByteStrings(bytestrings.take(last) :+ bytestrings(last).dropRight1(restToDropRight), length - n)
```
I realized `bytestrings.take(last)` can be `bytestrings.dropRight(last)`.
```
-      else if (restToDropRight == 0) new ByteStrings(bytestrings.take(last + 1), length - n)
-      else new ByteStrings(bytestrings.take(last) :+ bytestrings(last).dropRight1(restToDropRight), length - n)
+      else if (restToDropRight == 0) new ByteStrings(bytestrings.dropRight(bytestrings.length - last - 1), length - n)
+      else new ByteStrings(bytestrings.dropRight(bytestrings.length - last) :+ bytestrings(last).dropRight1(restToDropRight), length - n)
```

This is because we can expect `dropRight(M)` is used to _drop_ elements from the right-hand side of the string where `M <= N/2` and `N` is the length of the string. If this is the case, it can be said that (but not always the case as `ByteString` has `Vector[ByteString1]` data) `dropRight(M)` is more efficient than `take(N - M)` because `M <= N - M` is satisfied.

monkey-mas committed:
Change private[this] to private

monkey-mas committed:
Rename go(...) and variables

monkey-mas committed:
Add tests and fix tests

monkey-mas committed:
Use dropRight(...) instead of take(...) for the best use-case

We can expect that dropRight(M) is used to _drop_ elements from the right-hand side of the string
where M <= N/2 and N is the length of the string.
If this is the case, it can be said that (but not always the case as ByteString has Vector[ByteString1] data)
dropRight(M) is more efficient than take(N - M) because `M <= N - M` is satisfied.

Additionally, as _fullDrops_ represents the number of elements to _dropRight(N)_, some modifications was done :)

monkey-mas commented:
```
[error] (akka-typed-experimental/pr-validation:test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 4721 s, completed Sep 25, 2016 4:36:27 PM
```
Umm, I think the error isn't related to my code but this happened twice in a row... :(

johanandren commented:
It needs a rebase.

monkey-mas committed:
Fix benchmark

jrudolph commented:
PLS BUILD

monkey-mas committed:
[review fix] fix takeRight condition & remove unnecessary tuple instance

ref.
https://github.com/akka/akka/pull/21564#discussion_r88112094
https://github.com/akka/akka/pull/21564#discussion_r88112263

patriknw commented:
PLS BUILD
